netlink-proto/src/framed.rs

```diff
@@ -1,45 +1,50 @@
 use bytes::{BufMut, BytesMut};
 use std::{
+    fmt::Debug,
     io,
+    marker::PhantomData,
     pin::Pin,
     slice,
     task::{Context, Poll},
 };
 
 use futures::{Sink, Stream};
 use log::error;
-use tokio_util::codec::{Decoder, Encoder};
 
-use crate::sys::{Socket, SocketAddr};
+use crate::{
+    codecs::NetlinkMessageCodec,
+    sys::{Socket, SocketAddr},
+};
+use netlink_packet_core::{NetlinkDeserializable, NetlinkMessage, NetlinkSerializable};
 
-pub struct NetlinkFramed<C> {
+pub struct NetlinkFramed<T, C> {
     socket: Socket,
-    codec: C,
+    msg_type: PhantomData<fn(T) -> T>, // invariant
```

Comment:
Do you mind explaining why you chose `fn(T) -> T` instead of just `T`, and what you mean by the "invariant" comment?

Reply:
https://doc.rust-lang.org/nomicon/phantom-data.html has a few words on the matter.

I think invariant seems a good choice until someone comes up with a proper usecase for using covariant or contravariant (which, afaict, is mostly about lifetimes). `*mut T` would be invariant too but wouldn't derive `Send` nor `Sync` - `fn(T) -> T` derives `Send` and `Sync` afaict.

I'll add a comment in the code for it.